### ISSUE TITLE
False positive. `cast` gives `not-a-mapping` error

### ISSUE THREAD

**Issue by JohnVillalovos** [pylint-dev/pylint#9181]
### Bug description

```python
import dataclasses
from typing import Any, cast


@dataclasses.dataclass(kw_only=True)
class Resource:
    name: str


@dataclasses.dataclass(kw_only=True)
class ResourceParent:
    resource: Resource

    def __post_init__(self) -> None:
        if isinstance(self.resource, dict):
            self.resource = Resource(**cast(dict[str, Any], self.resource))
```

### Command used

```shell
pylint reproducer.py
```

### Pylint output

```python
************* Module scheduler_power_saver.reproducer
reproducer.py:16:39: E1134: Non-mapping value cast(dict[str, Any], self.resource) is used in a mapping context (not-a-mapping)
```

### Expected behavior

No error

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.6 (main, Oct  3 2023, 00:00:00) [GCC 13.2.1 20230728 (Red Hat 13.2.1-1)]
```

### OS / Environment

Fedora 38

**Comment by ziesemer** [comment]
https://github.com/pylint-dev/pylint/issues/4534 is similar, but for `E1101:no-member`.
